File: iOS/AppFactory/AppFactory/Http/Requester/GalleryRequester.swift
//
//  GalleryRequester.swift
//  AppFactory
//
//  Created by Leapfrog-Software on 2018/04/08.
//  Copyright © 2018年 Leapfrog-Inc. All rights reserved.
//

import Foundation

struct GalleryData {
    
    let id: String
    let name: String
    let engineerId: String
    let engineerArea: String
    let engineerOrganization: OrganizationType
    let cost: Int
    let evaluate: Int
    
    init?(data: Dictionary<String, Any>) {

        guard let id = data["id"] as? String else {
            return nil
        }
        self.id = id
        
        self.name = data["name"] as? String ?? ""
        self.engineerId = data["engineerId"] as? String ?? ""
        self.engineerArea = data["engineerArea"] as? String ?? ""
        self.engineerOrganization = OrganizationType.create(with: data["engineerOrganization"] as? Int ?? 0)
        self.cost = data["cost"] as? Int ?? 0
        self.evaluate = data["evaluate"] as? Int ?? 0
    }
}

struct GalleryResponseData {
    let page: Int
    let total: Int
    let galleryList: [GalleryData]
    
    init?(data: Dictionary<String, Any>) {
        
        guard let page = data["page"] as? Int else {
            return nil
        }
        self.page = page
        
        guard let total = data["total"] as? Int else {
            return nil
        }
        self.total = total
        
        self.galleryList = (data["galleryList"] as? Array<Dictionary<String, Any>>)?.flatMap { GalleryData(data: $0) } ?? []
    }
}

class GalleryRequester {
    
    class func get(sort: GallerySortType, page: Int, completion: @escaping ((GalleryResponseData?) -> ())) {
        
        var url = Constants.ServerRootUrl + Constants.ServerApiFile
        url += "?"
        url += ("command=getgalleries")
        url += "&"
        url += ("sort=" + sort.toString())
        url += "&"
        url += ("page=" + "\(page)")
        
        ApiManager.get(url: url, completion: { result, data in
            if result, let dic = data {
                completion(GalleryResponseData(data: dic))
            } else {
                completion(nil)
            }
        })
    }
}

File: Android/app/src/main/java/leapfrog_inc/appfactory/Http/Requester/GalleryRequester.java
package leapfrog_inc.appfactory.Http.Requester;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import leapfrog_inc.appfactory.Function.Constants;
import leapfrog_inc.appfactory.Http.Enum.GallerySortType;
import leapfrog_inc.appfactory.Http.Enum.OrganizationType;
import leapfrog_inc.appfactory.Http.HttpManager;

/**
 * Created by Leapfrog-Software on 2018/04/16.
 */

public class GalleryRequester {

    public static class GalleryData {

        public String id;
        public String name;
        public String engineerId;
        public String engineerName;
        public String engineerArea;
        public OrganizationType engineerOrganization;
        public int cost;
        public int evaluate;

        static public GalleryData create(JSONObject json) {

            try {
                String id = json.getString("id");
                String name = json.getString("name");
                String engineerId = json.getString("engineerId");
                String engineerName = json.getString("engineerName");
                String engineerArea = json.getString("engineerArea");
                OrganizationType organization = OrganizationType.create(json.getInt("engineerOrganization"));
                int cost = json.getInt("cost");
                int evaluate = json.getInt("evaluate");

                GalleryData engineerData = new GalleryData();
                engineerData.id = id;
                engineerData.name = name;
                engineerData.engineerId = engineerId;
                engineerData.engineerName = engineerName;
                engineerData.engineerArea = engineerArea;
                engineerData.engineerOrganization = organization;
                engineerData.cost = cost;
                engineerData.evaluate = evaluate;

                return engineerData;

            } catch (Exception e) {
            }

            return null;
        }
    }

    public static class GalleryResponseData {

        public int page;
        public int total;
        public ArrayList<GalleryData> galleryList = new ArrayList<GalleryData>();

        static public GalleryResponseData create(JSONObject json) {

            try {
                GalleryResponseData data = new GalleryResponseData();
                data.page = json.getInt("page");
                data.total = json.getInt("total");

                JSONArray galleries = json.getJSONArray("galleryList");
                for (int i = 0; i < galleries.length(); i++) {
                    GalleryData galleryData = GalleryData.create(galleries.getJSONObject(i));
                    if (galleryData != null) {
                        data.galleryList.add(galleryData);
                    }
                }
                return data;

            } catch (Exception e) {}

            return null;
        }
    }

    public static void get(GallerySortType sortType, int page, final GalleryRequesterCallback callback){

        HttpManager httpManager = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {

                if (result) {
                    try {
                        GalleryResponseData responseData = GalleryResponseData.create(new JSONObject(data));
                        if (responseData != null) {
                            callback.didReceiveData(true, responseData);
                            return;
                        }
                    } catch (Exception e) {}
                }
                callback.didReceiveData(false, null);
            }
        });

        StringBuffer url = new StringBuffer();
        url.append(Constants.ServerApiUrl);
        url.append("?");
        url.append("command=getgalleries");
        url.append("&");
        url.append(("sort=" + sortType.toValue()));
        url.append("&");
        url.append(("page=") + String.valueOf(page));

        httpManager.execute(url.toString(), "GET", null);
    }

    public interface GalleryRequesterCallback {
        void didReceiveData(boolean result, GalleryResponseData response);
    }
}
